From e247ad27c7f31bf17b4c565617bc0a1fe84f95d0 Mon Sep 17 00:00:00 2001
From: Yonit Halperin <yhalperi@redhat.com>
Date: Tue, 28 Feb 2012 06:11:46 +0100
Subject: [PATCH 03/12] spice: support ipv6 channel address in monitor events
 and in spice info

RH-Author: Yonit Halperin <yhalperi@redhat.com>
Message-id: <1330409507-6938-2-git-send-email-yhalperi@redhat.com>
Patchwork-id: 37655
O-Subject: [RHEL 6.3 qemu-kvm PATCH v2 1/2] spice: support ipv6 channel address in monitor events and in spice info
Bugzilla: 769512
RH-Acked-by: Alon Levy <alevy@redhat.com>
RH-Acked-by: Gerd Hoffmann <kraxel@redhat.com>
RH-Acked-by: Hans de Goede <hdegoede@redhat.com>
RH-Acked-by: Uri Lublin <uril@redhat.com>

RHBZ #769512

upstream: http://patchwork.ozlabs.org/patch/140147/

CC: Gerd Hoffmann <kraxel@redhat.com>
Signed-off-by: Yonit Halperin <yhalperi@redhat.com>
---
 ui/spice-core.c |   44 ++++++++++++++++++++++++++++++++++++++------
 1 files changed, 38 insertions(+), 6 deletions(-)

Signed-off-by: Michal Novotny <minovotn@redhat.com>
---
 ui/spice-core.c |   44 ++++++++++++++++++++++++++++++++++++++------
 1 files changed, 38 insertions(+), 6 deletions(-)

diff --git a/ui/spice-core.c b/ui/spice-core.c
index 297b7fa..196e6ab 100644
--- a/ui/spice-core.c
+++ b/ui/spice-core.c
@@ -171,7 +171,7 @@ static int channel_list_del(SpiceChannelEventInfo *info)
     return 0;
 }
 
-static void add_addr_info(QDict *dict, struct sockaddr *addr, int len)
+static void do_add_addr_info(QDict *dict, struct sockaddr *addr, int len)
 {
     char host[NI_MAXHOST], port[NI_MAXSERV];
     const char *family;
@@ -185,6 +185,38 @@ static void add_addr_info(QDict *dict, struct sockaddr *addr, int len)
     qdict_put(dict, "family", qstring_from_str(family));
 }
 
+static void add_addr_info(QDict *dict, SpiceChannelEventInfo *info,
+                          int is_client)
+{
+    struct sockaddr *addr;
+    int addr_len;
+
+#ifdef SPICE_CHANNEL_EVENT_FLAG_ADDR_EXT
+    if (info->flags & SPICE_CHANNEL_EVENT_FLAG_ADDR_EXT) {
+        if (is_client) {
+            addr = (struct sockaddr *)&info->paddr_ext;
+            addr_len = info->plen_ext;
+        } else {
+            addr = (struct sockaddr *)&info->laddr_ext;
+            addr_len = info->llen_ext;
+        }
+    } else {
+        fprintf(stderr, "spice: %s, extended address is expected\n",
+                        __func__);
+#endif
+        if (is_client) {
+            addr = &info->paddr;
+            addr_len = info->plen;
+        } else {
+            addr = &info->laddr;
+            addr_len = info->llen;
+        }
+#ifdef SPICE_CHANNEL_EVENT_FLAG_ADDR_EXT
+    }
+#endif
+    do_add_addr_info(dict, addr, addr_len);
+}
+
 static void add_channel_info(QDict *dict, SpiceChannelEventInfo *info)
 {
     int tls = info->flags & SPICE_CHANNEL_EVENT_FLAG_TLS;
@@ -204,7 +236,7 @@ static QList *channel_list_get(void)
     list = qlist_new();
     QTAILQ_FOREACH(item, &channel_list, link) {
         dict = qdict_new();
-        add_addr_info(dict, &item->info->paddr, item->info->plen);
+        add_addr_info(dict, item->info, true);
         add_channel_info(dict, item->info);
         qlist_append(list, dict);
     }
@@ -217,10 +249,10 @@ static void redhat_channel_event(int qevent, SpiceChannelEventInfo *info)
     QObject *data;
 
     client = qdict_new();
-    add_addr_info(client, &info->paddr, info->plen);
+    add_addr_info(client, info, true);
 
     server = qdict_new();
-    add_addr_info(server, &info->laddr, info->llen);
+    add_addr_info(server, info, false);
     qdict_put(server, "auth", qstring_from_str(auth));
 
     data = qobject_from_jsonf("{ 'client': %p, 'server': %p }",
@@ -254,10 +286,10 @@ static void channel_event(int event, SpiceChannelEventInfo *info)
     }
 
     client = qdict_new();
-    add_addr_info(client, &info->paddr, info->plen);
+    add_addr_info(client, info, true);
 
     server = qdict_new();
-    add_addr_info(server, &info->laddr, info->llen);
+    add_addr_info(server, info, false);
 
     if (event == SPICE_CHANNEL_EVENT_INITIALIZED) {
         qdict_put(server, "auth", qstring_from_str(auth));
-- 
1.7.7.6

